src/Http/Http/src/StreamResponseBodyFeature.cs

```diff
@@ -136,6 +136,23 @@ public virtual async Task CompleteAsync()
         {
             await _pipeWriter.CompleteAsync();
         }
+
+        if (PriorFeature != null)
+        {
+            await PriorFeature.CompleteAsync();
+        }
+    }
+
+    /// <summary>
+    /// Completes the current PipeWriter, if it exists
+    /// </summary>
+    /// <returns></returns>
+    public async Task CompleteWriterAsync()
+    {
+        if (_pipeWriter != null)
+        {
+            await _pipeWriter.CompleteAsync();
```

Comment:
Why Complete rather than just Flush?

Reply:
In case someone else has already completed it

Reply:
Should we just make this IAsyncDisposable and have this be the `DisposeAsync()` implementation? I think I like it because others might not find it obvious why you'd call `CompleteWriterAsync()`, but this would always be a safe disposal operation.

Reply:
Check if our other middleware using StreamResponseBodyFeature have the same bug.

Also what about this logic?
https://github.com/dotnet/aspnetcore/blob/543ed1bf3b7a0bbf52da5054315252218c2e2a0f/src/Http/Http/src/Internal/DefaultHttpResponse.cs#L69-L76

Reply:
I was thinking about that. The only thing is I wouldn't want to block there. I guess we can fire-and-forget. At least with Kestrel, ProduceEnd/WriteSuffix will ensure the flush gets awaited.

Reply:
eek

Reply:
I don't like it either @davidfowl. Do you think we should just stick with not auto-flushing the remaining bytes in the BodyWriter (if-any)?

I don't think most developers replacing `HttpResponse.Body` will know to do this. Then again, I don't think it's common to leave unflushed bytes in the `BodyWriter.`

@wtgodbe @JamesNK Does Grpc.AspNetCore leave unflushed bytes in the `BodyWriter` or does it call `PipeWriter/IHttpResponseBodyFeature.CompleteAsync()`?

Reply:
When gRPC serializes a single message (e.g. unary, client streaming) then it leaves unflushed bytes in the BodyWriter and lets ASP.NET Core flush them.

When gRPC serializes messages in a stream (e.g. server streaming, bidi streaming) then it explicitly flushes BodyWriter.

Reply:
What's the reason gRPC leaving unflushed bytes in the single message case rather than calling `CompleteAsync()`? If gRPC could be updated to call `CompleteAsync()`, it would make not flushing the `BodyWriter` when `HttpResponse.Body` is replaced less problematic.

Reply:
Can't we do something where we add an OnCompleted and flush the bytes or is that too late?

Reply:
> What's the reason gRPC leaving unflushed bytes in the single message case rather than calling `CompleteAsync()`? If gRPC could be updated to call `CompleteAsync()`, it would make not flushing the `BodyWriter` when `HttpResponse.Body` is replaced less problematic.

Probably performance. Allows the headers, body and response trailers to be sent together.

Reply:
> Can't we do something where we add an OnCompleted and flush the bytes or is that too late?

It's too late. At that point the HttpLoggingMiddleware has put the IHttpResponseBodyFeature back to it's original value.

>

Reply:
> Probably performance. Allows the headers, body and response trailers to be sent together.

This should also be achieved by calling `CompleteAsync()` without flushing first. This will call through to `Http2OutputProducer.WriteStreamSuffixAsync()` just like `HttpProtocol.ProduceEnd()` would.

Reply:
If we come up with a good solution for HttpResponse.Body then HttpLoggingMiddleware probably doesn't need to use StreamResponseBodyFeature at all.

Reply:
@JamesNK could gRPC use @halter73's `CompleteAsync` suggestion above instead of leaving unflushed bytes? This would allow us to come up with a better long-term fix in HttpResponse.Body rather than this intermediate step (plus that could be backported to 6.0, whereas this has API change)

Reply:
> If we come up with a good solution for HttpResponse.Body then HttpLoggingMiddleware probably doesn't need to use StreamResponseBodyFeature at all.

The problem here is that when we have a `PipeWriter` wrapping an arbitrary response body `Stream`, the wrapper has no choice other than to just write/flush the underlying stream (and optionally disposes the underlying stream) in `CompleteAsync()` instead of combining the flush with the final write to avoid empty DATA frames with the END_STREAM flag. There's no way to get the behavior gRPC wants unless we have access to the underlying `PipeWriter` or `IHttpResponseBodyFeature` and write data using the underlying `PipeWriter` directly so we can append data to the real response body without flushing it prematurely.

Reply:
@JamesNK any thoughts on updating gRPC as per Stephen's suggestion above?